### ISSUE TITLE
Allow quarkus-oidc to customize response body of 403

### ISSUE THREAD

**Issue by looorent** [quarkusio/quarkus#5751]
**Description**
When using quarkus-oidc, providing a wrong token (expired, wrong signature, realm, ...) returns a forbidden status, as expected. However, the response body is always empty.

In order to migrate an existing API (which already defines its own response bodies), we would require from quarkus-oidc the possibility to customize and give more information in the response body.

For example, an expired token could produce this response body:
```
{
   "reason": "token_expired",
   "message": "Your token is expired."
}
```

**Comment by sberyozkin** [comment]
This is a problem that we have to tackle somehow, the first problem is that we need to be able to launch an OIDC authentication from a pre-match JAX-RS chain so as to get the exception mappers utilized. The smaller problem would be to make `OIDCException` informative enough...
@stuartwdouglas Hi Stuart, what do you think ? The same issue was reported at the pure smallrye-jwt path, the users would like to customize the exceptions. 
Perhaps we can introduce quarkus specific exception mappers which would work across the security layer ?

**Comment by loicmathieu** [comment]
@sberyozkin the issue has already been discuss but not yet implemented :)

When we implements JSON representation for resteasy 404 (inside `io.quarkus.resteasy.runtime.NotFoundExceptionMapper`) and for undertow/Vet.x 500, I raise the fact that there is also exception mappers for authentification issues, but we didn't implement JSON representation for them. This was a leftover.

So we need to provide a default JSON reprensentation (close to the one we did for 404 and 500) inside `io.quarkus.resteasy.runtime.ForbiddenExceptionMapper` and `io.quarkus.resteasy.runtime.UnauthorizedExceptionMapper`.

A user can still override these ones by providing it's own Exception mapper as their priority is lower than the default one.

**Comment by sberyozkin** [comment]
@loicmathieu Hi Loic, not sure how much of that we can apply in this case, when the token has been validated and the roles extracted then yes, if 403 is reported is enforced in the JAX-RS chain we can map catch it, but in this case a user would like to catch the exceptions at the OIDC token verification time which is happening before the JAX-RS chain has started...

**Comment by loicmathieu** [comment]
@sberyozkin maybe we need more that what I propose, but what I propose must be done :)

It's hard to provide exception mappers for all possible exception generated by Quarkus. 

Reading the OIDC code, I saw an `OIDCException` that directly extends `RuntimeException`. According to me it's wrong as if OIDC is down the user should be denied access to the resource so we might throw a `ForbidenException` and log an error as it's forbidden because we cannot validate the token (not because the token is invalid). If not, the final exception mapper inside Vert.x/undertow will display the exception as all other ones.

**Comment by sberyozkin** [comment]
Hey @loicmathieu looks like you are proposing a solution for #5750 :-)

**Comment by stuartwdouglas** [comment]
OIDC does not depend on JAX-RS, this is a higher level API that can be applied to all HTTP implementations (Vert.x, Servlet etc). 

It sounds like what this really needs is a way to modify the response in the HTTP layer, as if the token is invalid then it may never actually hit the JAX-RS layer.

**Comment by sberyozkin** [comment]
@stuartwdouglas Would having an our own injectable `SecurityExceptionMapper` CDI bean typed by the base exception which `ForbiddenException`/etc extend do it ?

**Comment by stuartwdouglas** [comment]
Not really, there might not even be an exception, that is more of a JAX-RS concept. At the web layer it just looks at if auth it required or not. attempts to authenticate and then decides what to send.

This is complicated by the fact that the response might be sent by either JAX-RS or the Vert.x layer, so it needs some thought.

**Comment by sberyozkin** [comment]
#8570 may be the way to go as the JAX-RS users also want to catch the security exceptions with the exception mappers, see #9591

**Comment by stuartwdouglas** [comment]
I am going to look into making this customisable via vert.x failure routes.

**Comment by sberyozkin** [comment]
Hi @stuartwdouglas Can you please have a quick look at supporting the failure routes when possible, as we have a few issues all revolving around customizing the security exception responses.
CC @MarcusBiel, please watch this issue

**Comment by sberyozkin** [comment]
See also #12959

**Comment by wizedaoMinh** [comment]
I need this feature too. Vote up

**Comment by sberyozkin** [comment]
@michalvavrik Hi Michal, I think as far as the immediate impact on the Quarkus security is concerned, this one will likely have the biggest one. So please add it the list of the sec issues are are looking at :-)

**Comment by michalvavrik** [comment]
@sberyozkin Thanks, I'm always looking for new list items :-) I'll have a look later this week.

**Comment by sberyozkin** [comment]
@michalvavrik Sounds good, take your time please :-)

**Comment by michalvavrik** [comment]
I spend most of this day looking for scenarios where quarkus-oidc returns forbidden status as the issue describes. Provided examples in the description - wrong token (expired, wrong signature, realm, ...) all result in authentication exceptions - depending on context AuthenticationFailedException, AuthenticationCompletionException, AuthenticationRedirectException. We already fixed it in 2.15/main for extensions using vertx-http - exception mappers / failure handler (in RESTEasy Rective proactive=true scenario) are able to customize response. I can't find a scenario where oidc layer directly returns forbidden.

As far as forbidden response body customization is concerned, there are 2 types of scenarios:

1. security checks are applied by JAX-RS layer (e.g. RestEasy Reactive EagerSecurityHandler) or quarkus security (e.g. standard security interceptors for RBAC annotations).
2. HTTP policy security checks

Exceptions mappers are already able to handle **1.** and only scenario where exceptions are thrown that results in 403 are **2.**. 

If no-one objects, I am starting to work on fixing **2.** and that fix should also close the issue.
